lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,350 @@
+use crate::gelf_fields::*;
+use bytes::{BufMut, BytesMut};
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::{log_schema, DataType},
+    event::Event,
+    event::LogEvent,
+    schema,
+};
+
+/// On GELF encoding behavior:
+///   Graylog has a relaxed parsing. They are much more lenient than the spec would
+///   suggest. We've elected to take a more strict approach to maintain backwards compatability
+///   in the event that we need to change the behavior to be more relaxed, so that prior versions
+///   of vector will still work.
+///   The exception is that if 'Additional fields' are found to be missing an underscore prefix and
+///   are otherwise valid field names, we prepend the underscore.
+
+/// Config used to build a `GelfSerializer`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct GelfSerializerConfig;
+
+impl GelfSerializerConfig {
+    /// Creates a new `GelfSerializerConfig`.
+    pub const fn new() -> Self {
+        Self
+    }
+
+    /// Build the `GelfSerializer` from this configuration.
+    pub fn build(&self) -> GelfSerializer {
+        GelfSerializer::new()
+    }
+
+    /// The data type of events that are accepted by `GelfSerializer`.
+    pub fn input_type() -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement() -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // JSON, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Serializer that converts an `Event` to bytes using the GELF format.
+/// Spec: https://docs.graylog.org/docs/gelf
+#[derive(Debug, Clone)]
+pub struct GelfSerializer {
+    valid_regex: Regex,
+    conformed_log: Option<LogEvent>,
+}
+
+impl GelfSerializer {
+    /// Creates a new `GelfSerializer`.
+    pub fn new() -> Self {
+        GelfSerializer {
+            valid_regex: Regex::new(r"^[\w\.\-]*$").unwrap(),
+            conformed_log: None,
+        }
+    }
+
+    /// Encode event and represent it as JSON value.
+    pub fn to_json_value(&self, event: Event) -> Result<serde_json::Value, serde_json::Error> {
+        match event {
+            Event::Log(log) => serde_json::to_value(&log),
+            // TODO Metric and Trace shouldn't be valid but I was not finding a way to construct a
+            // custom serde_json::Error on the fly
+            Event::Metric(metric) => serde_json::to_value(&metric),
+            Event::Trace(trace) => serde_json::to_value(&trace),
```

Comment:
Want to make sure we handle this properly before merging!

Reply:
yes... looking into it as we speak!

Reply:
🤦 I meant to block merging with my review too, but I clicked the wrong thing - whoops